### DISCUSSION TITLE
Use form builder inside custom page (Filament Admin)

### DISCUSSION THREAD

**Discussion by ghost** [filamentphp/filament#1169] [Help] [answered]
Hello again! I'm new to Livewire and Filament.

I'm building a custom page for my site settings for admins. I'm a little bit lost on how do I leverage the custom form builder and spit it out on the view. I tried `{{ $this->form }}` but it doesn't seem to work. Probably I miss something.

On my `Pages/Settings.php` I'm not sure if I will use `getFormSchema()` or the `form()` method. 

```php
<?php

namespace App\Filament\Pages;

use Filament\Forms;
use Filament\Pages\Page;

class Settings extends Page
{
    protected static ?string $navigationIcon = 'heroicon-o-document-text';

    protected static string $view = 'filament.pages.settings';

    public $name;

    protected $rules = [
        'site_name' => 'required'
    ];

    // I'M NOT SURE IF I WILL USE THIS OR THE form() method
    protected function getFormSchema(): array 
    {
        return [
            Forms\Components\TextInput::make('name')->required(),
        ];
    } 

    public function submit()
    {
        $this->validate();

        // SAVE THE SETTINGS HERE
    }
}
```

Inside my view `settings.blade.php`

```php
<x-filament::page>
    <form wire:submit.prevent="submit">
        {{ $this->form }}
    
        <button type="submit">Save</button>
    
    </form>
</x-filament::page>
```

Got this error

```
 Property [$form] not found on component: [app.filament.pages.settings] 
```

Thanks and more power to Filament.

**Comment by ghost** [comment, marked as answer]
OMG it worked. I implemented `HasForms` and use `InteractsWithForms` inside Settings

```php
<?php

namespace App\Filament\Pages;

use Filament\Forms;
use Filament\Forms\Concerns\InteractsWithForms;
use Filament\Forms\Contracts\HasForms;
use Filament\Pages\Page;
use Filament\Resources\Form;

class Settings extends Page implements HasForms
{
    use InteractsWithForms;

    protected static ?string $navigationIcon = 'heroicon-o-document-text';

    protected static string $view = 'filament.pages.settings';

    public $site_name;

    protected $rules = [
        'site_name' => 'required'
    ];
    
    protected function getFormSchema()
    {
        return [
            Forms\Components\TextInput::make('site_name'),
        ];
    }

    public function submit()
    {
        $this->validate();

        // SAVE THE SETTINGS HERE
    }
}
```

I'm still digging through this just let me know if this is correct. The form is now rendering.

**Reply by danharrin** [reply]
Yes, that's exactly right! Make sure to call `$this->form->fill()` inside the `mount()` method though, this is imperative to ensure that the form works correctly.

**Reply by bilogic** [reply]
is this still how it should be in v3? I'm getting a blank page

**Reply by danharrin** [reply]
Follow the "adding a form to a Livewire component" instructions, treating the page and its view as the Livewire component.

**Reply by bilogic** [reply]
thanks, how do we populate a `logo` file upload field with the existing file?

```PHP
    public function mount(): void
    {
        $this->form->fill([
            'logo' => '...???', // what should I fill here? I sent in an array, and exception says string expected
        ]);
    }
```

**Reply by danharrin** [reply]
The path to the image that was returned the last time you saved. like `logo1234.png` or `directory/logo1234.png` if you have a `directory()` configured

**Reply by bilogic** [reply]
Hmm, I tried both `public/logo1234.png` and `/logo1234.png`, but the component is still unpopulated.

**Reply by danharrin** [reply]
probably no trailing slash or public/ required. Check what was returned when you uploaded the file in the first place. also check console errors that might suggest what is wrong with the URL.

**Reply by bilogic** [reply]
Hmm, no errors at all, actually I don't even see a request for the logo in Chrome Network tab.

I tried to examine my `Staff` resource that uses `SpatieMediaLibraryFileUpload` by adding `mount()` to `EditStaff.php`, the image show up as `null`.

```PHP
    public function mount(string|int $record): void
    {
        $record = Staff::find($record);
        dd($record);
        $this->form->fill($this->record->toArray());
    }
```

**Reply by danharrin** [reply]
You shouldn't need to override anything for a resource, it should work out of the box. Are you using Spatie media library? You didn't mention previously.

**Reply by bilogic** [reply]
Yes I'm using `SpatieMediaLibraryFileUpload`, sorry I'm very new to filament. Yes, for a resource, it works out of the box. But I'm trying to get it to work on a custom page.

**Reply by danharrin** [reply]
You don't need to fill anything if you're using `SpatieMediaLibraryFileUpload`. We do it for you, as long as you pass the `->model($record)` to the form as in our docs.

**Reply by bilogic** [reply]
Hmm, there are no models, it is a page where I would just like to upload some files and fill in some fields. My trouble is when try to edit/update this page. 

Here is a gist of the page trimmed down to its minimum. I'm having trouble on how to `form->fill()` it.
https://gist.github.com/bilogic/439e7634124a6aa9626c3e24236a9796

**Reply by danharrin** [reply]
Spatie Media Library requires a model to attach files to.

**Reply by danharrin** [reply]
`Staff::find($record)` by the looks of it

**Reply by bilogic** [reply]
ok, does the `FileUpload` also need a model? Or can it do without one? I tried both actually, but no joy.

**Reply by danharrin** [reply]
FileUpload doesn't need a model as it just returns a plain URL.

**Reply by bilogic** [reply]
> FileUpload doesn't need a model as it just returns a plain URL.

Ok, updated my code here https://gist.github.com/bilogic/439e7634124a6aa9626c3e24236a9796
For whatever reason, there are no error console logs nor any attempt to load the image file at the URL

**Reply by danharrin** [reply]
When you upload an image, what is returned from `$this->form->getState()` on submit

**Reply by bilogic** [reply]
I did a `dd($this->form->getState())` and got just the filename string
![image](https://github.com/filamentphp/filament/assets/946010/8a93544f-5de7-48a5-b1dc-ff247f24feb2)

I then did this
```PHP
public function mount(): void
{
    $this->form->fill([
        'logo' => 'file.png', // where should the path of this file be?
    ]);

    dd($this->data); // output below
    // array:1 [▼ // app/Filament/Pages/Test.php:28
    // "logo" => []
    // ]
    // why is logo = []? because file.png was not found? which folder should it be?
}
```

**Reply by bilogic** [reply]
1. I figured out the path was `storage/app/livewire-tmp/`
2. So, I created `storage/app/livewire-tmp/file.png`, but still faced `"logo" => []` when `'logo' => 'file.png'`

**Reply by danharrin** [reply]
Should be 

```
$this->form->fill([
        'logo' => 'bt9bW......png',
    ]);
```

In `storage/app/public`

**Reply by bilogic** [reply]
yea, i just figured that out too. thanks!

**Reply by bilogic** [reply]
Sorry, last question. Is there a way built-in way to make uploaded files require authentication instead of being accessible by anyone who has the link?

I saw https://github.com/filamentphp/filament/discussions/6080 

What does `->visibility('private')` do actually? The docs say https://filamentphp.com/docs/3.x/tables/columns/image#private-images temporary URLs but nothing more, when and how do they expire?

**Reply by danharrin** [reply]
When you use a real storage disk like S3, `visibility('private')` will add the correct permissions and generate temporary URLs that are only valid when you load the page initially

**Comment by Uhasith** [comment]
How can i get the route of a custom page? This route is not working. I want to make NavItem active when user goto that page?

This is the namespace of my custom page -> namespace App\Filament\Developer\Resources\UserResource\Pages;

```php
    ->navigationItems([

                NavigationItem::make('dashboard')
                    ->url(fn (): string => UserResource::getUrl('create-api'))
                    ->isActiveWhen(fn () => request()->routeIs('filament.developer.users.create.api'))
                    ->sort(3)
                    ->label('API'),

            ]);
``` 
@danharrin

**Reply by leandrocfe** [reply]
Don't tag @ Dan please.
Why did you create a resource page? You can create a custom page and this will display in the main menu.
 https://filamentphp.com/docs/3.x/panels/pages/#creating-a-page

**Comment by vlados** [comment]
I think it should be:
```blade
<x-filament-panels::page>
    <x-filament-panels::form wire:submit="submit">
        {{ $this->form }}
        <div>
        <x-filament::button type="submit" size="sm">
            Submit
        </x-filament::button>
        </div>
    </x-filament-panels::form>
</x-filament-panels::page>
```

**Reply by JamieBradders** [reply]
Thank you for sharing this @vlados. This combined with `getFormSchema()` in the `Settings.php` file worked for me 👍
